include/boost/url/any_allocator.hpp

```diff
@@ -0,0 +1,126 @@
+//
+// Copyright (c) 2022 Alan Freitas (alandefreitas@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/CPPAlliance/url
+//
+
+#ifndef BOOST_URL_ANY_ALLOCATOR_HPP
+#define BOOST_URL_ANY_ALLOCATOR_HPP
+
+#include <boost/url/detail/config.hpp>
+#include <boost/url/detail/any_allocator.hpp>
+
+#include <memory>
+#include <type_traits>
+
+namespace boost {
+namespace urls {
+
+/** A type-erased allocator with shared ownership.
+
+    This type satisfies the requirements for
+   <em>Allocator</em>
+
+    @par Specification
+    @li <a
+     href="https://en.cppreference.com/w/cpp/named_req/Allocator"
+       >Allocator (cppreference.com)</a>
+
+ */
+#ifdef BOOST_URL_DOCS
+template <class T>
+using any_allocator = __see_below__;
+#else
+template <class T>
+class any_allocator
+    : private detail::any_allocator_base
+{
+    std::shared_ptr<base> p_;
+
+    template <class U>
+    friend class any_allocator;
+
+public:
+    using value_type = T;
+
+    using size_type = boost::
+        allocator_size_type_t<any_allocator<T>>;
+
+    using difference_type =
+        allocator_difference_type_t<
+            any_allocator<T>>;
+
+    using reference = value_type&;
+
+    using const_reference = value_type const&;
+
+    using pointer = value_type*;
+
+    using const_pointer = value_type const*;
+
+    template <class U>
+    struct rebind
+    {
+        using other = any_allocator<U>;
+    };
+
+    using is_always_equal = std::false_type;
+
+#ifndef BOOST_URL_NO_GCC_4_2_WORKAROUND
+    any_allocator() = default;
+#endif
+
+    any_allocator(any_allocator const&) = default;
+
+    template <class U>
+    explicit any_allocator(
+        any_allocator<U> const& other) noexcept;
+
+    template <class Allocator>
+    explicit any_allocator(Allocator const& a);
+
+    any_allocator&
+    operator=(any_allocator const&) = default;
+
+    T*
+    allocate(std::size_t n) const;
+
+    void
+    deallocate(T* p, std::size_t n);
+
+    template <class U, class... Args>
+    void
+    construct(U* p, Args&&... args);
+
+    template< class U >
+    void destroy( U* p );
+
+    friend bool
+    operator==(const any_allocator& a,
+        const any_allocator& b) noexcept
+    {
+        return (a.p_ == b.p_) ||
+            (a.p_ != nullptr
+                && b.p_ != nullptr &&
+                (a.p_->type_id == b.p_->type_id &&
+                    a.p_->is_equal(*b.p_)));
+    }
+
+    friend bool
+    operator!=(const any_allocator& a,
+        const any_allocator& b) noexcept
+    {
+        return !(a == b);
+    }
+};
+#endif
+
+} // namespace urls
+} // namespace boost
```

Comment:
I don't bother putting the word "namespace" because it doesn't add any information, but if you feel better with it there, I won't complain. See:
https://github.com/boostorg/beast/blob/5bdd1efb8be740bb5425adc054c990818134f3e2/include/boost/beast/http/basic_parser.hpp#L691

Reply:
So can I leave the word "namespace"? That would be one less thing to get wrong because it's how it's generated automatically.

Reply:
you can leave it in if you want, but I think you will over time learn that these "automatic" formatting tools are going to cause you more work than they relieve.